contracts/src/token_locker.cairo

```diff
@@ -0,0 +1,177 @@
+
+#[starknet::interface]
+trait ITokenLocker<TContractState> {
+    fn lock(ref self: TContractState, token_contract: ContractAddress, amount: u256);
+    fn unlock(ref self: TContractState, token_contract: ContractAddress, lock_time: u64);
+    fn get_locked_amount(
+        self: @TContractState,
+        token_contract: ContractAddress,
+        owner: ContractAddress,
+        lock_time: u64
+    ) -> u256;
+    fn get_time_left(
+        self: @TContractState,
+        token_contract: ContractAddress,
+        owner: ContractAddress,
+        lock_time: u64
+    ) -> u64;
+}
+
+#[starknet::contract]
+mod TokenLocker {
+    use openzeppelin::token::erc20::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait};
+    use starknet::{ContractAddress, get_caller_address, get_contract_address, get_block_timestamp};
+    use unruggable::errors::{
+        ZERO_LOCK, ZERO_LOCK_TIME, LOCK_EXIST, TRANSFER_FAIL, LOCK_NONEXIST, STILL_LOCKED
+    };
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        TokenLocked: TokenLocked,
+        TokenUnlocked: TokenUnlocked
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TokenLocked {
+        locker: ContractAddress,
+        token: ContractAddress,
+        lock_time: u64,
+        amount: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TokenUnlocked {
+        unlocker: ContractAddress,
+        token: ContractAddress,
+        unlock_time: u64,
+        amount: u256
+    }
+
+    #[storage]
+    struct Storage {
+        lock_time: u64,
+        locks: LegacyMap<(ContractAddress, ContractAddress, u64), u256>,
+    }
+
+    /// Constructor called once when the contract is deployed.
+    /// # Arguments
+    /// * `lock_time` - Locking period as timestamp.
+    #[constructor]
+    fn constructor(ref self: ContractState, lock_time: u64,) {
+        self.lock_time.write(lock_time);
+    }
+
+    #[abi(embed_v0)]
+    impl TokenLocker of super::ITokenLocker<ContractState> {
+        /// Locks tokens from callers account
+        /// # Arguments
+        /// * `token_contract` - Address of token contract that is going to be locked
+        /// * `amount` - Amount of tokens that is going to be locked
+        fn lock(ref self: ContractState, token_contract: ContractAddress, amount: u256) {
+            assert(amount != 0, ZERO_LOCK);
+
+            let caller = get_caller_address();
+            let current_time = get_block_timestamp();
+            assert(self.locks.read((token_contract, caller, current_time)) == 0, LOCK_EXIST);
+
+            self.locks.write((token_contract, caller, current_time), amount);
+
+            let this_address = get_contract_address();
+            let initial_balance = ERC20ABIDispatcher { contract_address: token_contract }
+                .balanceOf(this_address);
+            ERC20ABIDispatcher { contract_address: token_contract }
+                .transferFrom(caller, this_address, amount);
+            assert(
```

Comment:
any scenario where the transfer could fail and the execution continue? I think this is an extra unnecessary step but need confirmation

Reply:
In case it is token locker. there may be token transfer which has transfer taxes so calculation can be wrong here. it is checking that

Reply:
hmm, so that means we can't lock tokens with transfer tax?